db/migrations/postgres/012_query_index.up.sql

```diff
@@ -0,0 +1,6 @@
+CREATE INDEX events_instance_id_started_at_idx ON events USING btree(instance_id, started_at);
+
+CREATE INDEX jobs_instance_id_idx ON jobs USING btree(instance_id);
+CREATE INDEX jobs_busy_instance_id_idx ON jobs USING btree(instance_id) WHERE claimed_at IS NOT NULL AND finished_at IS NULL;
+CREATE INDEX jobs_queue_order_idx ON jobs USING btree((-1 * priority), scheduled_at, submitted_at) WHERE finished_at IS NULL AND claimed_at IS NULL;
+CREATE INDEX jobs_instance_id_unfinished_idx ON jobs USING btree(instance_id) WHERE finished_at IS NULL;
```

Comment:
How much of an advantage do we get with these partial indexes `jobs_busy_instance_id_idx` and `jobs_instance_id_unfinished_idx` on `instance_id` when there is also a complete index?

I'm not familiar with how these work in postgres. Does it use only the partial index and ignore the complete index if it recognises it can do so?

Reply:
>I'm not familiar with how these work in postgres. Does it use only the partial index and ignore the complete index if it recognises it can do so?

In theory the query planner picks the index most appropriate to the current query.
The partial clauses match things from the WHERE condition of the query.

> How much of an advantage do we get with these partial indexes jobs_busy_instance_id_idx and jobs_instance_id_unfinished_idx on instance_id when there is also a complete index?

The more specific the index, the better the query performance. (However maintaining each extra index has a performance impact).


I've extracted them from the common queries:
- jobs_instance_id_idx matches a specific instance status query
- jobs_busy_instance_id_idx matches part of a nested query
- jobs_queue_order_idx matches the sorting constraints on a large query
- jobs_instance_id_unfinished_idx matches part of a query looking for what is next

I'm not certain these are the perfect indexes, but they should help.
We can run some analysis after deploying to see which the query planner decides to use.

Reply:
thanks